tpm2/tpm2_test.go

```diff
@@ -0,0 +1,318 @@
+	}
+
+	keyparms := RSAParams{
+		AlgRSA,
+		AlgSHA1,
+		0x00030072,
+		[]byte(nil),
+		AlgAES,
+		128,
+		AlgCFB,
+		AlgNull,
+		0,
+		2048,
+		uint32(0x00010001),
+		[]byte(nil),
+	}
+	privateBlob, publicBlob, err := CreateKey(rw, parentHandle, []int{7}, "", "01020304", keyparms)
+	if err != nil {
+		t.Fatalf("CreateKey failed: %s", err)
+	}
+
+	keyHandle, _, err := Load(rw, parentHandle, "", "", publicBlob, privateBlob)
+	if err != nil {
+		t.Fatalf("Load failed: %s", err)
+	}
+
+	_, name, _, err := ReadPublic(rw, keyHandle)
+	if err != nil {
+		t.Fatalf("ReadPublic failed: %s", err)
+	}
+
+	// Generate Credential
+	credential := []byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 0xa, 0xb, 0xc, 0xd, 0xe, 0xf, 0x10}
+	credBlob, encryptedSecret0, err := MakeCredential(rw, parentHandle, credential, name)
+	if err != nil {
+		FlushContext(rw, keyHandle)
+		FlushContext(rw, parentHandle)
+		t.Fatalf("MakeCredential failed: %s", err)
+	}
+
+	recoveredCredential1, err := ActivateCredential(rw, keyHandle, parentHandle, "01020304", "", credBlob, encryptedSecret0)
+	if err != nil {
+		FlushContext(rw, keyHandle)
+		FlushContext(rw, parentHandle)
+		t.Fatalf("ActivateCredential failed: %s", err)
+	}
+	if bytes.Compare(credential, recoveredCredential1) != 0 {
+		FlushContext(rw, keyHandle)
+		FlushContext(rw, parentHandle)
+		t.Fatalf("Credential and recovered credential differ: got %v, want %v", recoveredCredential1, credential)
+	}
+
+	FlushContext(rw, keyHandle)
+}
+
+func TestCombinedContextTest(t *testing.T) {
+	if !*runIntegration {
+		t.SkipNow()
+	}
+	rw, err := OpenTPM("/dev/tpm0")
+	if err != nil {
+		t.Fatalf("OpenTPM failed: %v", err)
+	}
+	defer rw.Close()
+
+	pcrs := []int{7}
+	keySize := 2048
+	quotePassword := ""
+
+	primaryparms := RSAParams{
+		AlgRSA,
+		AlgSHA1,
+		FlagStorageDefault,
+		[]byte(nil),
+		AlgAES,
+		128,
+		AlgCFB,
+		AlgNull,
+		0,
+		uint16(keySize),
+		uint32(0x00010001),
+		[]byte(nil),
+	}
+	rootHandle, _, err := CreatePrimary(rw, HandleOwner, pcrs, "", "", primaryparms)
+	if err != nil {
+		t.Fatalf("CreatePrimary failed: %v", err)
+	}
+	defer FlushContext(rw, rootHandle)
+
+	// CreateKey (Quote Key)
+	keyparms := RSAParams{
+		AlgRSA,
+		AlgSHA1,
+		FlagSignerDefault,
+		[]byte(nil),
+		AlgNull,
+		0,
+		AlgECB,
+		AlgRSASSA,
+		AlgSHA1,
+		uint16(keySize),
+		uint32(0x00010001),
+		[]byte(nil),
+	}
+	quotePrivate, quotePublic, err := CreateKey(rw, rootHandle, pcrs, "", quotePassword, keyparms)
+	if err != nil {
+		t.Fatalf("CreateKey failed: %v", err)
+	}
+
+	quoteHandle, _, err := Load(rw, rootHandle, "", quotePassword, quotePublic, quotePrivate)
+	if err != nil {
+		t.Fatalf("Load failed: %v", err)
+	}
+	defer FlushContext(rw, quoteHandle)
+
+	saveArea, err := ContextSave(rw, quoteHandle)
+	if err != nil {
+		t.Fatalf("ContextSave failed: %v", err)
+	}
+	FlushContext(rw, quoteHandle)
+
+	quoteHandle, err = ContextLoad(rw, saveArea)
+	if err != nil {
+		t.Fatalf("Load failed: %v", err)
+	}
+
+	FlushContext(rw, quoteHandle)
+}
```

Comment:
Not in this PR, but in future ones, please add some negative tests that show that the TPM fails appropriately, e.g., when the password auth is wrong.

Reply:
Acknowledged.